gax-grpc/src/main/java/com/google/api/gax/grpc/TransportProvider.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2017, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.BetaApi;
+import java.io.IOException;
+import java.util.concurrent.ScheduledExecutorService;
+
+/** The settings used to create a Transport. */
+@BetaApi
+public interface TransportProvider {
+
+  /** True if the Transport needs an executor. */
+  boolean needsExecutor();
+
+  /**
+   * Provides a Transport, which could either be a new instance for every call, or the same
+   * instance, depending on the implementation.
+   *
+   * <p>This method should only be called if {@link #needsExecutor()} returns false.
+   */
+  Transport getContext() throws IOException;
+
+  /**
+   * Provides a Transport, which could either be a new instance for every call, or the same
+   * instance, depending on the implementation.
+   *
+   * <p>This method should only be called if {@link #needsExecutor()} returns true.
+   */
+  Transport getContext(ScheduledExecutorService executor) throws IOException;
```

Comment:
I was trying to syntactically separate those cases, because someone might think they need to always provide a ScheduledExecutorService, but that might consume heavy resources.